### PR TITLE
feat(options): add support for `--mode` flag (`options.mode`)

### LINKED ISSUE #225: No way to set mode with no-config webpack 4
There currently seems to be no way to set the `mode` property Webpack [requires](https://webpack.js.org/concepts/mode/) when not using a `webpack.config.js` file. As of version 4, webpack [doesn't require any configuration](https://webpack.js.org/guides/getting-started/#using-a-configuration), but, if no configuration file is present, `webpack-serve` always works in production mode and produces a warning.

* Operating System: Windows
* Node Version: 10.1.0
* NPM Version: 5.5.1
* webpack Version: 4.16.0
* webpack-serve Version: 2.0.2

This issue is for a:

- [ ] **bug**
- [X] **feature** request
- [ ] **modification** request

##### CLI Command

```console
  $ webpack-serve
```

##### webpack.config.js
`webpack.config.js` *does not exist*

### Expected Behavior
You should be able to set `mode` without having to create a configuration file with the sole purpose of setting the mode.

### Actual Behavior
There is no way to set the mode apart from creating a configuration file. When running `webpack-serve`, you always get a production build with the following warning printed in the console:

```
WARNING in configuration
The 'mode' option has not been set, webpack will fallback to 'production' for this value. Set 'mode' option to 'development' or 'production' to enable defaults for each environment.
You can also set it to 'none' to disable any default behavior. Learn more: https://webpack.js.org/concepts/mode/
```

### How Do We Reproduce?
Follow the Webpack [Getting Started](https://webpack.js.org/guides/getting-started/) guide up until [Creating a Bundle](https://webpack.js.org/guides/getting-started/#creating-a-bundle). Then install the `webpack-serve` package and run it.

### New Feature Use Case

One possible solution is to add a `--mode` argument like `webpack-cli`. Another option could be to use an environment variable or to default `mode` to either `production` or `development` when no config file is present (`development` seems to make the most sense).

### PR DESCRIPTION
**Pull request by cilliemalan**
This PR contains a:

- [ ] **bugfix**
- [x] new **feature**
- [ ] **code refactor**
- [ ] **test update** 
- [ ] **typo fix**
- [ ] **metadata update**

### Motivation / Use-Case
This PR resolves #225.

There is currently no way to specify `mode` when run against a project utilizing [Webpack 4 zero-configuration](https://webpack.js.org/guides/getting-started/#using-a-configuration). If webpack-serve is run on a project that doesn't have any configuration, it will always run in production mode and produce a warning.

This PR adds support for a `--mode` command line argument or `mode` option that allows one to specify the mode in which Webpack will compile (development, production, or none).

### Additional Info
If `--mode` is specified and a configuration file *is* present, the value specified will override what is configured. This is the same behaiour as `webpack-cli`.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -67,6 +67,8 @@ Options
   --log-time       Instruct the logger for webpack-serve and dependencies to display a timestamp
   --hmr            Specify whether or not the client should apply Hot Module Replacement patches (default: true)
   --reload         Specify whether or not the middleware should reload the page for build errors (default: true)
+  --mode           Enable production optimizations or development hints.
+                   Choices: development, production, none
```
Comment (shellscape) [inline README.md]: Please change this to: `Modes: `

File: lib/config.js [hunk 2]
```diff
@@ -119,6 +120,11 @@ module.exports = {
 
       found = [].concat(found).map((conf) => prepare(conf));
 
+      if (mode) {
```
Comment (shellscape) [inline lib/config.js]: not the right spot to do this. the flag value should be applied to the `result` in the `apply()` method of `lib/flags.js`. it's impossible to know which config in a multi-config that a user intends to apply a flag to, so it should apply to all configs.

File: schemas/flags.js [hunk 3]
```diff
@@ -14,7 +14,8 @@ module.exports = {
     type: 'string',
   },
   'dev-ware': {
-    desc: 'Set options for webpack-dev-middleware. e.g. --dev-ware.publicPath /',
+    desc:
+      'Set options for webpack-dev-middleware. e.g. --dev-ware.publicPath /',
```
Comment (shellscape) [inline schemas/flags.js]: please keep an eye on your editor making unnecessary style changes

File: schemas/flags.js [hunk 4]
```diff
@@ -70,6 +71,11 @@ Use {dim --no-hot-client} to disable webpack-hot-client`,
     desc: chalk`Specify whether or not the middleware should reload the page for build errors {dim (default: true)}`,
     type: 'boolean',
   },
+  mode: {
+    desc: chalk`Enable production optimizations or development hints.
+{dim Choices: development, production, none}`,
```
Comment (shellscape) [inline schemas/flags.js]: `Modes: ` please

File: test/config.test.js [hunk 5]
```diff
@@ -122,4 +122,20 @@ describe('config', () => {
...
+    const argv = { mode: 'development' };
+    const config = undefined;
+    return load(argv, { config }).then((configs) => {
+      expect(configs).toMatchSnapshot();
+    });
+  });
+
+  test('load: mode overrides config', () => {
+    const argv = { mode: 'production' };
+    const config = './test/fixtures/webpack-4-defaults/webpack.config.js';
```
Comment (shellscape) [inline test/config.test.js]: this is a good test, but we also need a test with the multi-config to make sure the mode was applied correctly to an array of configs

**Review by shellscape — CHANGES_REQUESTED** [review]
great start, see comments

File: schemas/flags.js (hunk 3 above)
Comment (cilliemalan) [reply schemas/flags.js]: I'll fix it but note that `lint-staged` is doing this.

**Comment by cilliemalan** [comment]
@shellscape I've made most of the changes you requested.

I've implemented the mode override in `apply` in `lib/options`. I don't quite see how it'll work in `lib/flags` as the config is not available there. Let me know if this is okay?

I've added tests for `options`, checking that the mode is set in single config and multiple config. I've also added a test fixture and two tests for Webpack 4 no-config (one normal one, and one checking that mode is set on the compiler when passed as an argument).

Let me know if this is okay!

**Comment by cilliemalan** [comment]
Fixed the test.

**Review by G-Rath — APPROVED** [review]
@shellscape Looks decent.

As cilliemalan said, I'm not sure how you'd implement the mode override in `lib/options`, unless maybe you also change `lib/compiler` to actually look for that property in the `options` argument...?

That's the only part I think you need to double check - the rest looks solid, and seems to properly implement all the changes you requested in your last review.

**Review by michael-ciniawsky — APPROVED** [review]

**Comment by alexander-akait** [comment]
DEPRECATED. Please use `webpack-dev-server` (he is in support and update mode again). Feel free to create a issue if some features are not implemented in `webpack-dev-server`. 

Why deprecated `webpack-serve` ?
- The author stopped developing the package.
- Two development servers are misleading for developers.
- Hard to maintain two packages with same purpose.
- Some `dependencies` are not available to fixes/implement new features/etc (https://www.npmjs.com/package/webpack-hot-client and https://github.com/shellscape/koa-static) and there's nothing we can do here.
- `webpack-dev-server` is stable.

Thanks for using `webpack`! We apologize for the inconvenience. In the future, we will avoid such situations.

**Pull request closed without merging**